### PR TITLE
Several updates

### PR DESCRIPTION
**Pull request by chuckha**
* Use machine name instead of generated name tied to kind
* Improve delete functionality to remove node from child cluster
* Set nodeRefs

Signed-off-by: Chuck Ha <chuckh@vmware.com>

### CODE REVIEW

File: actuators/machine.go [hunk 1]
```diff
@@ -78,14 +79,25 @@ func (m *Machine) Create(ctx context.Context, c *clusterv1.Cluster, machine *clu
...
-			controlPlaneNode, err := actions.AddControlPlane(c.Name, machine.Spec.Versions.ControlPlane)
+			controlPlaneNode, err := actions.AddControlPlane(c.Name, machine.GetName(), machine.Spec.Versions.ControlPlane)
 			if err != nil {
 				fmt.Printf("%+v", err)
 				return err
 			}
-			name := providerID(controlPlaneNode.Name())
-			machine.Spec.ProviderID = &name
-			return m.save(old, machine)
+			nodeUID, err := actions.GetNodeRefUID(c.GetName(), controlPlaneNode.Name())
```
Comment (pablochacin) [inline actuators/machine.go]: Isn't this same code repeated below? maybe use a helper function? also will reduce the overall `create` function length.

File: actuators/machine.go [hunk 2]
```diff
@@ -132,18 +155,43 @@ func (m *Machine) Create(ctx context.Context, c *clusterv1.Cluster, machine *clu
...
+		Name:       worker.Name(),
+		UID:        types.UID(nodeUID),
+	}
+	return m.save(old, machine, nodeRef)
 }
 
+// Delete returns nil when the machine no longer exists or when a successful delete has happened.
 func (m *Machine) Delete(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
-	return actions.DeleteNode(cluster.Name, providerNameToLookupID(*machine.Spec.ProviderID))
+	if machine.Status.NodeRef != nil {
```
Comment (pablochacin) [inline actuators/machine.go]: Maybe add the machine name to reflect with machine the messagesbelow refers to?

File: kind/actions/cluster_actions.go [hunk 3]
```diff
@@ -274,28 +274,63 @@ func SetNodeRef(clusterName, nodeName string) error {
...
+		"--kubeconfig", "/etc/kubernetes/admin.conf",
+		"delete", "node", nodeName,
 	)
 	lines, err := exec.CombinedOutputLines(cmd)
 	if err != nil {
 		for _, line := range lines {
 			fmt.Println(line)
 		}
-		return errors.Wrap(err, "failed to remove node from cluster")
+		return errors.Wrap(err, "failed update providerID")
```
Comment (pablochacin) [inline kind/actions/cluster_actions.go]: I think the correct message is "failed to remove node from cluster" as this code is in `DeleteClusterNode` function

**Comment by sethp-nr** [comment]
The changes look reasonable enough to me so far, but I'm running into some trouble trying this out:

```
$ capdctl setup
Error: error loading config: decoding failure: no kind "Cluster" is registered for version "kind.sigs.k8s.io/v1alpha3" in scheme "sigs.k8s.io/kind/pkg/cluster/config/encoding/scheme.go:34"
panic: exit status 1

goroutine 1 [running]:
main.makeManagementCluster(0x1713884, 0x4)
	/Users/spellegrino/Code/Go/src/github.com/kubernetes-sigs/cluster-api-provider-docker/cmd/capdctl/main.go:186 +0x188
main.main()
	/Users/spellegrino/Code/Go/src/github.com/kubernetes-sigs/cluster-api-provider-docker/cmd/capdctl/main.go:82 +0x6cb
```

I'll see if I can figure that out and get back to you. Also, I found a place in the README the rename missed: https://github.com/kubernetes-sigs/cluster-api-provider-docker/blame/master/README.md#L30-L32

File: actuators/machine.go (hunk 2 above)
Comment (chuckha) [reply actuators/machine.go]: I actually don't think this message is necessary, I'm going to nix it entirely

File: actuators/machine.go (hunk 1 above)
Comment (chuckha) [reply actuators/machine.go]: agreed, I'll see if I can extract this

File: kind/actions/cluster_actions.go (hunk 3 above)
Comment (chuckha) [reply kind/actions/cluster_actions.go]: very correct, good eyes

**Comment by chuckha** [comment]
@sethp-nr ah yeah, I'm also actively developing on the :latest tag, so I need to move over to dev and re-push that back. I'd recommend you build your own image at the moment

**Comment by sethp-nr** [comment]
Yeah, that's what I ended up doing – it looks like the README references a script that I didn't find. I've got another PR I'm about to send you for a couple quick things.

File: kind/actions/cluster_actions.go [hunk 4]
```diff
@@ -306,15 +306,19 @@ func GetNodeRefUID(clusterName, nodeName string) (string, error) {
 	return strings.TrimSpace(lines[0]), nil
 }
 
+// DeleteClusterNode will remove
```
Comment (sethp-nr) [inline kind/actions/cluster_actions.go]: I think this is missing a few more words

**Comment by chuckha** [comment]
> I haven't tried multiple control planes / deleting one yet, but otherwise everything looks great and is working like I expect.

sweet, I'm going to fix up the delete comment and go ahead and merge it, thanks for the feedback @pablochacin and @sethp-nr !

The multi-control plane one needs a timeout of 15 minutes on an XL instance on AWS

**Comment by vincepri** [comment]
/lgtm
/hold

Feel free to unhold when ready

**Comment by chuckha** [comment]
/hold cancel
/lgtm

image

**Pull request merged**